homeassistant/components/knx/project.py

```diff
@@ -0,0 +1,120 @@
+"""Handle KNX project data."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+from typing import Final
+
+from xknx.dpt import DPTBase
+from xknxproject import XKNXProj
+from xknxproject.models import (
+    Device,
+    GroupAddress as GroupAddressModel,
+    KNXProject as KNXProjectModel,
+    ProjectInfo,
+)
+
+from homeassistant.components.file_upload import process_uploaded_file
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.storage import Store
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STORAGE_VERSION: Final = 1
+STORAGE_KEY: Final = f"{DOMAIN}/knx_project.json"
+
+
+@dataclass
+class GroupAddressInfo:
+    """Group address info for runtime usage."""
+
+    address: str
+    name: str
+    description: str
+    dpt_main: int | None
+    dpt_sub: int | None
+    transcoder: type[DPTBase] | None
+
+
+class KNXProject:
+    """Manage KNX project data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+    ) -> None:
+        """Initialize project data."""
+        self.hass = hass
+        self._store = Store[KNXProjectModel](hass, STORAGE_VERSION, STORAGE_KEY)
+
+        self.loaded: bool
+        self.devices: dict[str, Device]
+        self.group_addresses: dict[str, GroupAddressInfo]
+        self.info: ProjectInfo | None
+
+        self.initial_state()
+
+    def initial_state(self) -> None:
+        """Set initial state for project data."""
+        self.loaded = False
+        self.devices = {}
+        self.group_addresses = {}
+        self.info = None
+
+    async def load_project(self, data: KNXProjectModel | None = None) -> None:
+        """Load project data from storage."""
+        if project := data or await self._store.async_load():
+            self.devices = project["devices"]
+            self.info = project["info"]
+
+            for _ga_model in project["group_addresses"].values():
+                ga_info = self._create_group_address_info(_ga_model)
+                self.group_addresses[ga_info.address] = ga_info
```

Comment:
I would probably have used a dict comprehension. Thou below did look rather ugly :) so feel free to ignore.

```python
self.group_addresses = {
  (ga_info := self._create_group_address_info(_ga_model)).address: ga_info
  for ga_model in project["group_addresses"].values():
]
```

Reply:
> so feel free to ignore

I will 🤣